### PR TITLE
Improve logging in DeltaScheduler

### LINKED ISSUE #8912: Track time ops are sitting in inbound queue before they are processed
This is break out from #8911 

It tracks # 3 item in the original list:
Track how long ops are sitting in inbound queue until it is processed (that would capture impact of DeltaScheduler & ScheduleManager).

Basically, track time from when "push" event is fired for an op (probably first op in a batch is sufficient) till we start or end of op processing. Maybe track both - time to start processing, and time to process? I think we track later somewhere.

As always - it should be sampled not to produce too much data.

### PR DESCRIPTION
**Pull request by NicholasCouri**
Fix for #8909  - Adding additional logging to the DeltaScheduler before invoking the inboundqueue.resume() inside batchEnd call 

Information Added to the new event **DeltaManagerPaused**
a) how many ops we already processed till that moment?
b) how many ops are remained in queue?
c) current numberOfTurns 
d) processing and elapsed times

Added to the existing one (InboundOpsProcessingTime): 
a)Once done (reached idle), how many ops actually were processed (will be different from # 2, as ops are coming in).

@vladsud - it seems we do need the isScheduling  as we can call batchEnd() multiple times in case we have multiple batches ?

### CODE REVIEW

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 1]
```diff
@@ -117,6 +132,8 @@ export class DeltaScheduler {
                 numberOfOps: this.schedulingLog.numberOfOps,
                 numberOfTurns: this.schedulingLog.numberOfTurns,
                 processingTime: this.schedulingLog.totalProcessingTime,
+                numberOfOpsBefore: this.queueLengthBeforeBatchBegin,
+                numberOfOpsRemaining:  this.deltaManager.inbound.length,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:136]: "idle" event is fired when queue is empty, so numberOfOpsRemaining is not interesting here as it should be always zero.

The most useful metric here is number of ops we processed in total - from very first beginBatch message till getting here. We would need to collect that data manually (somehow), as looking (alone) at inbound queue size at various times is not helpful - ops could be added to inbound queue as processing happens.

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 2]
```diff
@@ -86,7 +90,18 @@ export class DeltaScheduler {
...
                 // eslint-disable-next-line @typescript-eslint/no-floating-promises
                 this.deltaManager.inbound.pause();
+
                 setTimeout(() => {
+                    if (this.schedulingLog) {
+                        this.logger.sendTelemetryEvent({
+                            eventName: "DeltaManagerPaused",
+                            duration: elapsedTime,
+                            numberOfOps: this.deltaManager.inbound.length,
+                            numberofOpsProcessed: this.queueLengthBeforeBatchBegin - this.deltaManager.inbound.length,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:100]: Similar as below - numberofOpsProcessed is not what it says, as inbound can be receiving ops as we go. It would be great to take a look at how we can capture actual number of ops processed.

I see couple ways:
- Add extra payload on "batchEnd" event that signifies number of ops processed
- batchEnd already has message payload. If we add message payload on beginBatch, we can capture sequence numbers and substract, getting number of ops processed
- we can listen on "op" method from container runtime and count (simpler, but a bit more expensive probably from CPU processing).

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ import {
 export class DeltaScheduler {
     private readonly deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>;
     // The time for processing ops in a single turn.
-    public static readonly processingTime = 20;
+    public static readonly processingTime = 50;
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:28]: Is this change in any way related to https://github.com/microsoft/FluidFramework/issues/8908? Not sure I understand the reason behind increasing this time

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 4]
```diff
@@ -117,6 +132,8 @@ export class DeltaScheduler {
                 numberOfOps: this.schedulingLog.numberOfOps,
                 numberOfTurns: this.schedulingLog.numberOfTurns,
                 processingTime: this.schedulingLog.totalProcessingTime,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:134]: If I read code correctly, processingTime is the time we spent processing ops (minus waiting due to timeouts injected by this code, but not including waiting due to unrelated code pausing op processing). It would be great to capture also total time from start of measurement till end of measurement, which would including amount of delays we injected here.

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 5]
```diff
@@ -129,6 +146,8 @@ export class DeltaScheduler {
             this.schedulingCount++;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:146]: Observation RE this.isScheduling & this.schedulingCount:

It feels to me that this.schedulingLog can be removed, as this.schedulingLog !== undefined can be used for same check. Make it a bit easier to follow code.

Also we do not count or report cases where it took us single batch to process all ops till idle. This skews data toward 2+ batches till idle. I think we should populate this.schedulingLog in batchEnd() even if queue is empty at that moment.

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 3 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:28]: yes, though we do not have enough data to directly tie these two things.

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 2 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:100]: I tried listening to ops. Let me know what you think.

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 5 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:146]: That should do it. I also moved the initialization of the schedulingLog to happen before the check to see if the queue is empty (shouldRunScheduler)

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 3 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:28]: It is related to #9505  where Vlad suggests the increase to 50 ms instead of 20 - Should I separate the check ins ?

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 6]
```diff
@@ -86,7 +90,18 @@ export class DeltaScheduler {
 
                 // eslint-disable-next-line @typescript-eslint/no-floating-promises
                 this.deltaManager.inbound.pause();
+
                 setTimeout(() => {
+                    if (this.schedulingLog) {
+                        this.logger.sendTelemetryEvent({
+                            eventName: "DeltaManagerPaused",
+                            duration: elapsedTime,
+                            numberOfOps: this.deltaManager.inbound.length,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:99]: I'd rename numberOfOps to something like opsRemainingToProcess
And for consistency - numberofOpsProcessed -> opsProcessed

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 4 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:134]: The processingTime during InboundOpsProcessingTime counts for everything. 
Just so I can understand, for this particular event, do you want to include the time it took from pausing until resume gets called?

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 6 above)
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:99]: We should add all state tracking for logging to`this.schedulingLog` or remove it and track these independently. Right now, its a mix and it adds confusion.
Instead of initializing the `schedulingLog` in `batchEnd` we should do it in `batchBegin` and assign the current inbound queue length to it. And then use it as a reference to log the number of ops processed, remaining, etc. whenever we want to log.

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 7]
```diff
@@ -43,6 +43,9 @@ export class DeltaScheduler {
     // to process the ops.
     private schedulingCount: number = 0;
 
+    // Keeps track of the Queue length before processing a batch.
+    private queueLengthBeforeBatchBegin: number = 0;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:47]: I think this should be called `numberOfOpsToProcess` or something similar since we are interested in ops.

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 8]
```diff
@@ -86,7 +90,18 @@ export class DeltaScheduler {
 
                 // eslint-disable-next-line @typescript-eslint/no-floating-promises
                 this.deltaManager.inbound.pause();
+
                 setTimeout(() => {
+                    if (this.schedulingLog) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:95]: Why is this inside setTimeout? Shouldn't this be logged before we call setTimeout?

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:99]: We don't use queueLengthBeforeBatchBegin if this.schedulingLog is undefined, so it would make sense for it to be part of the schedulingLog itself.

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:134]: I am a little unclear here. My understanding is that there are 2 kind of delays this code ends up introducing:
1. Time taken to pause the queue and then resume it.
2. Time spend by other code that uses the JS thread when this code releases it.
Do you mean to say we measure both these times?

Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts]: My goal was to collect the elapsedTime (before we pause)  and then , before we resume, when the timeout lambda gets called. But I'm not sure if it will be useful - while debugging it seems to be :)

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 6 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts]: yes, I removed queueLengthBeforeBatchBegin - the firstSequenceNumber and lastSequenceNumbers would give us the number of ops processed, isn't it ?

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 7 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts]: removed it.

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 8 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:133]: The idea was to be able to get info of when the resume happens....

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 9]
```diff
@@ -25,81 +29,111 @@ import {
...
+    private firstSequenceNumber: number = 0;
+
     constructor(
         deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>,
         private readonly logger: ITelemetryLogger,
     ) {
         this.deltaManager = deltaManager;
         this.deltaManager.inbound.on("idle", () => { this.inboundQueueIdle(); });
+        this.deltaManager.inbound.on("op", (op: ISequencedDocumentMessage) => {
+            if (op.type === MessageType.Operation) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:64]: I don't think this is correct. For each op, this handler will probably be called before batchBegin which is called when container runtime processes the op (after delta manager). So, batchProcessBegin will be false even though this may be a batch.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:64]: I think it would be better to add the op as a parameter to batchBegin and beginEnd. ScheduleManager can pass in the op when it calls those methods.

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:133]: I see. `waitingToResume` is also something we want to log so that makes sense :)

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 10]
```diff
@@ -25,81 +29,111 @@ import {
...
+                    // queue, yield the thread and schedule a resume.
+
+                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                    this.deltaManager.inbound.pause();
+
+                    setTimeout(() => {
+                        if (this.schedulingLog) {
+                            this.logger.sendTelemetryEvent({
+                                eventName: "DeltaManagerPaused",
+                                duration: TelemetryLogger.formatTick(elapsedTime),
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:114]: The names for various times logged here are not clear.
- Do we need duration aka elaspedTime? That is the time spent in processing the last batch not the total time spent in processing a set of ops until we ran out of time.
- I think logging this.schedulingLog.totalProcessingTime should be enough. Although, we should update it to add elaspedTime before we log this message.
- waitingToResume -> timeSpentWhenPaused or timeToResume? Basically, indicate that this is a time.
Also, we should add `ms` suffix to all time related properties. That will make it easier to understand when looking at telemetry.

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 11]
```diff
@@ -25,81 +29,111 @@ import {
...
+                    // We have processed ops for more than the total processing time. So, pause the
+                    // queue, yield the thread and schedule a resume.
+
+                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                    this.deltaManager.inbound.pause();
+
+                    setTimeout(() => {
+                        if (this.schedulingLog) {
+                            this.logger.sendTelemetryEvent({
+                                eventName: "DeltaManagerPaused",
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:113]: I think this should be renamed to `InboundOpsProcessingTime` and we should remove the one in idle state. The one logged in idle state will just be a duplicate of this one now that we will fire this on batchEnd right?

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 12]
```diff
@@ -25,81 +29,111 @@ import {
...
+                    this.processingStartTime = undefined;
+                    // Increase the total processing time. Keep doing this after each turn until all the ops have
+                    // been processed. This way we keep the responsiveness at the beginning while also making sure
+                    // that all the ops process fairly quickly.
+                    this.totalProcessingTime += this.processingTimeIncrement;
+
+                    // If we are logging the telemetry this time, update the telemetry log object.
+                    if (this.schedulingLog) {
+                        this.schedulingLog.numberOfTurns++;
+                        this.schedulingLog.totalProcessingTime += elapsedTime;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:135]: This should be done before logging the event above so it captures this data.

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 13]
```diff
@@ -101,6 +108,25 @@ export class DeltaScheduler {
...
                 }
+
+                setTimeout(() => {
+                    if (this.schedulingLog) {
+                        this.logger.sendTelemetryEvent({
+                            eventName: "InboundOpsPartialProcessingTime",
+                            duration: TelemetryLogger.formatTick(elapsedTime),
+                            opsProcessed: this.schedulingLog.lastSequenceNumber -
+                                this.schedulingLog.firstSequenceNumber,
+                            opsRemainingToProcess: this.deltaManager.inbound.length,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:119]: opsRemainingToProcess should be always zero here, as "idle" event is fired by queue only when there are no more elements in the queue. I'd remove it from logging here.

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 14]
```diff
@@ -109,25 +135,27 @@ export class DeltaScheduler {
...
+            this.schedulingLog.totalProcessingTime += currentTime - this.processingStartTime!;
 
             this.logger.sendTelemetryEvent({
                 eventName: "InboundOpsProcessingTime",
-                numberOfOps: this.schedulingLog.numberOfOps,
+                opsRemainingToProcess: this.schedulingLog.opsRemainingToProcess,
                 numberOfTurns: this.schedulingLog.numberOfTurns,
-                processingTime: this.schedulingLog.totalProcessingTime,
+                processingTime: TelemetryLogger.formatTick(this.schedulingLog.totalProcessingTime),
+                opsProcessed: this.schedulingLog.lastSequenceNumber - this.schedulingLog.firstSequenceNumber,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:147]: opsProcessed - please add 1 to it, length = last - first + 1

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 15]
```diff
@@ -101,6 +108,25 @@ export class DeltaScheduler {
                     this.schedulingLog.numberOfTurns++;
                     this.schedulingLog.totalProcessingTime += elapsedTime;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:109]: I'd recommend naming these two variables differently, i.e. find better name (fort first one) - it's confusing to have exactly same names with quite different meaning / semantics:
this.totalProcessingTime
this.schedulingLog.totalProcessingTime

**Review by vladsud — APPROVED** [review]

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 16]
```diff
@@ -57,41 +59,46 @@ export class DeltaScheduler {
         this.deltaManager.inbound.on("idle", () => { this.inboundQueueIdle(); });
     }
 
-    public batchBegin() {
+    public batchBegin(message: ISequencedDocumentMessage) {
         if (!this.processingStartTime) {
             this.processingStartTime = performance.now();
         }
+        if (this.schedulingLog === undefined && this.schedulingCount % 2000 === 0) {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:66]: One thing to change here: I'd change it to === 1
Reason for that - we are not capturing very first 2000 batches, which means we have zero visibility into what happens at the beginning of container lifetime. I expect most really large batches of ops to be on boot, but this query shows zero hits for "loading" phase:

union Office_Fluid_FluidRuntime_*
| where Data_eventName contains "InboundOpsProcessingTime"
| summarize count() by Data_containerLifecycleState

I'd say - given that we are operating on number of idles / batches here, we can reduce it to % 500 or so - that would result in number of hits closer to number of times we record fluid:telemetry:OpPerf:ConnectionSpeed, which makes data a bit easier to analyze (less skew toward start or end of the session).

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 17]
```diff
@@ -109,25 +135,27 @@ export class DeltaScheduler {
...
             // Add the time taken for processing the final ops to the total processing time in the
             // telemetry log object.
+            const currentTime = performance.now();
             // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-            this.schedulingLog.totalProcessingTime += performance.now() - this.processingStartTime!;
+            this.schedulingLog.totalProcessingTime += currentTime - this.processingStartTime!;
 
             this.logger.sendTelemetryEvent({
                 eventName: "InboundOpsProcessingTime",
-                numberOfOps: this.schedulingLog.numberOfOps,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:117]: Might be useful to capture here this.schedulingCount. This allows us to see various skews, i.e. see if start of the session sees more pausing than middle of the session.

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 13 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:119]: This is not on idle event but before we resume the inbound queue. I was assuming this could be non-zero here. Isn't it the case ?

Comment (vladsud) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:119]: Yes, fine here, comment should have been for InboundOpsProcessingTime event :) There it should be always zero

File: packages/runtime/container-runtime/src/deltaScheduler.ts (hunk 15 above)
Comment (NicholasCouri) [reply packages/runtime/container-runtime/src/deltaScheduler.ts:109]: Great point - changed the this.totalProcessingTime to currentAllowedProcessingTimeForTurn

File: packages/runtime/container-runtime/src/deltaScheduler.ts [hunk 18]
```diff
@@ -57,50 +59,72 @@ export class DeltaScheduler {
         this.deltaManager.inbound.on("idle", () => { this.inboundQueueIdle(); });
     }
 
-    public batchBegin() {
+    public batchBegin(message: ISequencedDocumentMessage) {
         if (!this.processingStartTime) {
             this.processingStartTime = performance.now();
         }
+        if (this.schedulingLog === undefined && this.schedulingCount % 500 === 0) {
+            // Every 2000th time we are scheduling the inbound queue, we log telemetry for the
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/deltaScheduler.ts:67]: nit: this should say 500th time now.
